src/emulator/storage/apis/firebase.ts

```diff
@@ -128,6 +130,48 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
     return res.json(new OutgoingFirebaseMetadata(metadata));
   });
 
+  firebaseStorageAPI.post(`/b/:bucketId/o/:objectId[(:)]generateSignedUrl`, async (req, res) => {
+    const timeToLive = req.body.ttl;
+    const msInAWeek = 7 * 24 * 60 * 60 * 1000;
+
+    if (timeToLive < 0 || timeToLive > msInAWeek) {
+      return res.status(400).json({
+        error: {
+          code: 400,
+          message: `Invalid TTL.`,
+        },
+      });
+    }
+    const currentDate = new Date()
+      .toISOString()
+      .replaceAll("-", "")
+      .replaceAll(":", "")
+      .replaceAll(".", "");
+
+    const unsignedUrl = `${req.protocol}://${req.hostname}:${req.socket.localPort}/v0/b/${req.params.bucketId}/o/${req.params.objectId}?alt=media&X-Firebase-Date=${currentDate}&X-Firebase-Expires=${timeToLive}`;
```

Comment:
I think we should split out express js details from url construction. 

Maybe create an object like `CreateSignedUrlRequest` similar to https://github.com/firebase/firebase-tools/blob/57dd7fef80d9b9294da41bb93cf25a1108674459/src/emulator/storage/files.ts#L90 and supply all the properties needed to generate signed urls. We could move this to a new file and add a helper:

```
func generateSignedUrl(req: CreateSignedUrlRequest): CreateSignedUrlResponse {
  ...
}
```

All input validation / parsing can remain here, but I'd move the `(timeToLive < 0 || timeToLive > msInAWeek)` inside this new helper. The helper should throw when out of range ttls are passed.

We can hardcode `https` instead of using `req.protocol`.

Instead of passing `hostname` and `port` separately, we can pass them as one parameter called `baseUrl` `${req.hostname}:${req.socket.localPort}`.

Reply:
I would also split out all the details related to computing the signature into another helper called by `generateSignedUrl(...)` (it can live in the same file). Specifically, this logic:

createHmac("sha256", privateKey).update(unsignedUrl).digest("base64");